File: Chapter 07/RoundButton/RoundButton.vb
'
' The Definitive Guide to the .NET Compact Framework
'   Apress : ISBN 1-59059-095-3 : 2003
'
' Authors:
'   Larry Roof
'     email: larry.roof@larryroof.com
'     web site: www.larryroof.com
'   Dan Fergus 
'     email: danf@forestsoftwaregroup.com
'     web site: www.forestsoftwaregroup.com
'
' You can obtain updates to the samples included with this title 
' at the following sites:
'   www.larryroof.com
'   www.forestsoftwaregroup.com
'   www.apress.com
'
'
' Copyright (c) 2003 by Larry Roof and Dan Fergus
'   
' The information and code provided hereunder (collectively referred to
' as "Software") is provided as is without warranty of any kind, either
' express or implied, including but not limited to the implied 
' warranties of merchantability and fitness for a particular purpose.
' In no event shall either Larry Roof and/or Dan Fergus be liable for any
' damages whatsoever including direct, indirect, incidental, consequential,
' loss of business profits or special damages, even if Larry Roof and/or Dan
' Fergus have been advised of the possibility of such damages. Some states
' do not allow the exclusion or limitation of liability for consequential
' or incidental damages so the foregoing limitation may not apply.
'
' This software, and the examples and techniques demonstrated within this
' software, may be incorporated into your software products subject to the
' following conditions:
'
' 1. This software cannot be distributed in its original form to any other
'    individuals for any purpose.
'
' 2. License to use this software is only provided to those individuals that
'    purchase a copy of the book "The Definitive Guide to the .NET Compact
'    Framework", published by Apress Publishing.
'
' 3. You agree to indemnify, hold harmless, and defend Larry Roof and Dan
'    Fergus from and against any claims or lawsuits, including attorneys 
'    fees, that arise or result from the use or distribution of your software 
'    product and any modifications to the Software.
'
Imports System.Drawing

Public Class RoundButton
    Inherits System.Windows.Forms.Control

    Dim _buttonDrawnDown As Boolean = False
    Dim _btnColor As Color = Color.LightGreen
    Dim _btnDisabledColor As Color = Color.DarkGreen

    Protected Overrides Sub OnPaint(ByVal e As System.Windows.Forms.PaintEventArgs)
        DrawButton(e.Graphics, Me.ForeColor, _btnColor)
        MyBase.OnPaint(e)
    End Sub

    Protected Overrides Sub OnMouseMove(ByVal e As System.Windows.Forms.MouseEventArgs)
        Try
            If e.Button = MouseButtons.Left Then
                If ((e.X > Me.ClientRectangle.Top And e.X < Me.ClientRectangle.Bottom) And (e.Y > Me.ClientRectangle.Left And e.Y < Me.ClientRectangle.Right)) Then
                    If (_buttonDrawnDown) Then
                        '   I'm inside the button face
                        Dim g As Graphics = Me.CreateGraphics()
                        DrawButton(g, Me.ForeColor, _btnColor)
                        g.Dispose()
                        _buttonDrawnDown = False
                    End If
                Else
                    '	Moving outside the button
                    If _buttonDrawnDown Then
                        Dim g As Graphics = Me.CreateGraphics()
                        DrawButton(g, Me.ForeColor, Color.Black)
                        g.Dispose()
                        _buttonDrawnDown = True
                    End If
                End If
            End If
            '           End If
            MyBase.OnMouseMove(e)
        Catch
        End Try
    End Sub

    Private Sub DrawButton(ByVal g As Graphics, ByVal fColor As Color, ByVal bColor As Color)
        '    use the button height as the diameter of the round end
        Dim bRadius As Int32 = Me.ClientRectangle.Height / 2
        '    Create a Brush object and initialize it to Nothinb
        Dim sb As Brush = Nothing

        If Me.Enabled Then
            sb = New SolidBrush(bColor)
        Else
            sb = New SolidBrush(_btnDisabledColor)
        End If

        '    Create variables to hold button size for calculations
        Dim aTop As Int32 = Me.ClientRectangle.Top
        Dim aLeft As Int32 = Me.ClientRectangle.Left
        Dim aWidth As Int32 = Me.ClientRectangle.Width
        Dim aHeight As Int32 = Me.ClientRectangle.Height

        '    First, draw the semi-circle that is the left end of the button
        g.FillEllipse(sb, aLeft, aTop, 2 * bRadius, 2 * bRadius)
        '    Now, draw the rectangle for the button body
        g.FillRectangle(sb, aLeft + bRadius, aTop, aWidth - (2 * bRadius), aHeight)
        '    Lastly, draw the semi-circle that is the right end of the button
        g.FillEllipse(sb, aWidth - (2 * bRadius), aTop, 2 * bRadius, 2 * bRadius)

        '    Now write the text to the button face
        '    Use the font set in the base control property
        Dim fnt As Font = New Font(Me.Font.Name, Me.Font.Size, Me.Font.Style)
        '    We measure the string to see if
        Dim siz As SizeF = g.MeasureString(Me.Text, fnt)
        '    Here we draw the string, centering it in the buttons client space
        '    Note that it is centered both horizontally and verticlaly in the space
        g.DrawString(Me.Text, fnt, New SolidBrush(fColor), _
            (aWidth - siz.Width) / 2, (aHeight - siz.Height) / 2)
        sb.Dispose()
    End Sub

    Protected Overrides Sub OnEnabledChanged(ByVal e As System.EventArgs)
        Dim g As Graphics = Me.CreateGraphics()
        If Me.Enabled Then
            DrawButton(g, Me.ForeColor, _btnColor)
        Else
            DrawButton(g, Me.ForeColor, _btnDisabledColor)
        End If
        g.Dispose()
        MyBase.OnEnabledChanged(e)
    End Sub

    Protected Overrides Sub OnMouseUp(ByVal e As System.Windows.Forms.MouseEventArgs)
        Try
            Dim g As Graphics = Me.CreateGraphics()
            DrawButton(g, Me.ForeColor, _btnColor)
            g.Dispose()
            _buttonDrawnDown = False
            MyBase.OnMouseUp(e)
        Catch
        End Try
    End Sub

    Protected Overrides Sub OnMouseDown(ByVal e As System.Windows.Forms.MouseEventArgs)
        Try
            Dim g As Graphics = Me.CreateGraphics()
            DrawButton(g, Color.White, Color.Black)
            g.Dispose()
            _buttonDrawnDown = True
            MyBase.OnMouseDown(e)
        Catch
        End Try
    End Sub

    Property ButtonColor() As Color
        Get
            Return _btnColor
        End Get
        Set(ByVal Value As Color)
            _btnColor = Value
        End Set
    End Property
    Property ButtonColorDisabled() As Color
        Get
            Return _btnDisabledColor
        End Get
        Set(ByVal Value As Color)
            _btnDisabledColor = Value
        End Set
    End Property

End Class

File: Appendix F/RoundButton.cs
using System;
//using System.Data;
using System.Drawing;
using System.Windows.Forms;

#if NETCFDESIGNTIME
[assembly: System.CF.Design.RuntimeAssemblyAttribute("DansButtonControl, Version=1.0.100.0, Culture=neutral, PublicKeyToken=null")]
#endif

namespace Apress.Windows.Controls
{
	public class RoundButton : System.Windows.Forms.Control, IDisposable
	{
		bool buttonDrawnDown = false;
		Color _btnColor = Color.LightGreen;
		Color _btnDisabledColor = Color.DarkGreen;

		public RoundButton()
		{
		}

		protected override void OnEnabledChanged(EventArgs e)
		{
			Graphics g = this.CreateGraphics();
			DrawButton(g, this.ForeColor, _btnColor);
			g.Dispose();
			base.OnEnabledChanged(e);
		}

		protected override void OnResize(EventArgs ea) 
		{
			this.Refresh();
		}

		protected override void OnPaint(PaintEventArgs e)
		{
			DrawButton(e.Graphics, this.ForeColor, _btnColor);
			base.OnPaint (e);
		}

		protected override void OnMouseMove(MouseEventArgs e)
		{
			try
			{
				if( e.Button == MouseButtons.Left )
				{
					if( (e.X < this.ClientRectangle.Top || e.X > this.ClientRectangle.Width) || (e.Y < this.ClientRectangle.Top || e.Y > this.ClientRectangle.Height) )
					{
						if( buttonDrawnDown )
						{
							Graphics g = this.CreateGraphics();
							DrawButton(g, this.ForeColor, _btnColor);
							g.Dispose();
							buttonDrawnDown = false;
						}
					}
					else
					{
						//	Moving inside the button
						if( !buttonDrawnDown )
						{
							Graphics g = this.CreateGraphics();
							DrawButton(g, Color.White /*this.ForeColor*/, Color.Black);
							g.Dispose();
							buttonDrawnDown = true;
						}
					}
				}
				base.OnMouseMove(e);
			}
			catch
			{
			}
		}

		protected override void OnMouseDown(MouseEventArgs e)
		{
			try
			{
				Graphics g = this.CreateGraphics();
				DrawButton(g, Color.White, Color.Black);
				g.Dispose();
				buttonDrawnDown = true;
				base.OnMouseDown(e);
			}
			catch
			{
			}
		}

		protected override void OnMouseUp(MouseEventArgs e)
		{
			try
			{
				Graphics g = this.CreateGraphics();
				Color sb;

				if( this.Enabled )
					sb = _btnColor;
				else
					sb = _btnDisabledColor;

				DrawButton(g, this.ForeColor, _btnColor);
				g.Dispose();
				buttonDrawnDown = false;
				this.Refresh();
				base.OnMouseUp(e);
			}
			catch
			{
			}
		}

		public void DrawButton(Graphics g, Color foreColor, Color backColor)
		{
			//	use the button height as the diameter of the round end
			int bRadius = this.ClientRectangle.Height / 2;
			Brush sb = null;

			sb = new SolidBrush(backColor);

			int aTop = this.ClientRectangle.Top;
			int aLeft = this.ClientRectangle.Left;
			int aWidth = this.ClientRectangle.Width;
			int aHeight = this.ClientRectangle.Height;

			
			g.FillEllipse(sb, aLeft, aTop, 2 * bRadius, 2 * bRadius);
			g.FillRectangle(sb, aLeft + bRadius, aTop, aWidth - (2 * bRadius), aHeight);
			g.FillEllipse(sb, aWidth - (2*bRadius), aTop, 2*bRadius, 2*bRadius);

			//	Now write the text to the button face
			Font fnt = new Font(this.Font.Name, this.Font.Size, this.Font.Style);
			SizeF siz = g.MeasureString(this.Text, fnt);
			g.DrawString(this.Text, fnt, new SolidBrush(foreColor), (aWidth - siz.Width)/2, (aHeight-siz.Height)/2);

			sb.Dispose();
		}

		#region IDisposable Members

		void System.IDisposable.Dispose()
		{
			// TODO:  Add RoundButton.System.IDisposable.Dispose implementation
		}

		#endregion

#if NETCFDESIGNTIME
	[
		System.ComponentModel.Category("Appearance"), System.ComponentModel.Description("The color for the Calendar caption font")
	]
#endif
		public Color ButtonColor
		{
			get
			{
				return _btnColor;
			}
			set
			{
				_btnColor = value;
			}
		}

#if NETCFDESIGNTIME
	[
		System.ComponentModel.Category("Appearance"), System.ComponentModel.Description("The color for the Calendar caption font")
	]
#endif
		public Color DisabledButtonColor
		{
			get
			{
				return _btnDisabledColor;
			}
			set
			{
				_btnDisabledColor = value;
			}
		}
	}
}
